File: android-tv/app/src/main/java/com/streamflow/tv/viewmodel/DetailViewModel.kt
package com.streamflow.tv.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.streamflow.tv.data.model.MovieDetail
import com.streamflow.tv.data.repository.MovieRepository
import com.streamflow.tv.data.repository.UserDataRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch

data class DetailUiState(
    val movie: MovieDetail? = null,
    val isLoading: Boolean = true,
    val error: String? = null,
    val isInMyList: Boolean = false,
    val selectedServer: String = ""
)

class DetailViewModel : ViewModel() {

    private val repository = MovieRepository()
    private val _uiState = MutableStateFlow(DetailUiState())
    val uiState: StateFlow<DetailUiState> = _uiState
    
    private var userDataRepository: UserDataRepository? = null

    fun loadMovie(slug: String, userRepo: UserDataRepository? = null) {
        if (userRepo != null) {
            this.userDataRepository = userRepo
        }
        
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                val movie = repository.getMovieDetail(slug)
                val isInMyList = userDataRepository?.isInMyList(slug) ?: false
                val firstServer = movie.episodes?.mapNotNull { it.serverName.ifBlank { null } }?.firstOrNull() ?: ""
                _uiState.value = _uiState.value.copy(
                    movie = movie, 
                    isLoading = false,
                    isInMyList = isInMyList,
                    selectedServer = firstServer
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Failed to load movie details"
                )
            }
        }
    }

    fun selectServer(server: String) {
        _uiState.value = _uiState.value.copy(selectedServer = server)
    }

    fun toggleMyList() {
        val movie = _uiState.value.movie ?: return
        val currentStatus = _uiState.value.isInMyList
        
        viewModelScope.launch {
            if (currentStatus) {
                userDataRepository?.removeFromMyList(movie.slug)
            } else {
                userDataRepository?.addToMyList(movie.toMovie())
            }
            _uiState.value = _uiState.value.copy(isInMyList = !currentStatus)
        }
    }
}

File: ios-app/KvNetflix/ViewModels/DetailViewModel.swift
import Foundation
import Combine

@MainActor
public final class DetailViewModel: ObservableObject {
    @Published public private(set) var movie: MovieDetail?
    @Published public private(set) var isLoading: Bool = true
    @Published public private(set) var error: String? = nil
    @Published public private(set) var isInMyList: Bool = false
    @Published public var selectedServer: String = ""
    @Published public private(set) var servers: [String] = []
    @Published public private(set) var relatedMovies: [Movie] = []

    private let movieRepo = MovieRepository.shared
    private let userRepo = UserDataRepository.shared

    public init() {}

    public var episodesForSelectedServer: [Episode] {
        guard let episodes = movie?.episodes else { return [] }
        if selectedServer.isEmpty {
            return episodes
        }
        let filtered = episodes.filter { $0.serverName == selectedServer }
        return filtered.isEmpty ? episodes : filtered
    }

    public func loadMovie(slug: String) {
        isLoading = true
        error = nil

        Task {
            do {
                let detail = try await movieRepo.getMovieDetail(slug: slug)
                self.movie = detail
                self.isInMyList = userRepo.isInMyList(slug: slug)

                let serverList = Array(Set(detail.episodes?.map { $0.serverName }.filter { !$0.isEmpty } ?? [])).sorted()
                self.servers = serverList
                self.selectedServer = serverList.first ?? ""
                self.isLoading = false

                // Load related / explore movies
                self.loadRelatedMovies(currentSlug: slug)
            } catch {
                self.isLoading = false
                self.error = error.localizedDescription
            }
        }
    }

    public func selectServer(_ server: String) {
        self.selectedServer = server
    }

    public func toggleMyList() {
        guard let movie = movie else { return }
        if isInMyList {
            userRepo.removeFromMyList(slug: movie.slug)
            isInMyList = false
        } else {
            userRepo.addToMyList(movie.toMovie())
            isInMyList = true
        }
    }

    private func loadRelatedMovies(currentSlug: String) {
        Task {
            do {
                let items = try await movieRepo.exploreMovies()
                self.relatedMovies = items.filter { $0.slug != currentSlug }
            } catch {
                print("[DetailViewModel] Error loading related movies: \(error)")
            }
        }
    }
}
